website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,651 @@
+to compare array and other BSON types.
+
+* Ascending sort uses the smallest element of any BSON type from the array.
+* Descending sort uses the largest element of any BSON type from the array.
+* Empty array is less than null.
+
+[Filtering](#comparison-in-filter-operator), commonly used in `$lt` and `$gt` operators,
+applies the following rules to compare array and other BSON types.
+
+* `$lt` uses the smallest element of the same BSON type as non-array from the array.
+* `$gt` uses the largest element of the same BSON type as non-array from the array.
+* Empty array should not match other BSON type.
+
+[Updating](#comparison-in-update-operator) with `$max` and `$min` operators applies the following rules
+to compare array and other BSON types.
+
+* For array and other BSON type comparison, BSON comparison order is used.
+* Empty array is greater than null using BSON comparison order.
+
+#### Example of Ascending and Descending Sort
+
+How does array compare with other BSON type using one of the array element?
+In this example, `2` and `[3, 1]` are compared using ascending and descending sort.
+
+##### Array and Other BSON Type Ascending Sort
+
+The following steps show the ascending sort comparison for `2` and `[3, 1]`:
+
+1. It gets the smallest element of array `[3, 1]` which is `1`.
+2. It compares `2` and `1`, `2` is greater than `1`.
+
+Therefore, when using ascending sort, `2` is greater than `[3, 1]`.
+
+##### Array and Other BSON Type Descending Sort
+
+The following steps show the descending sort comparison for  `2` and `[3, 1]`.
+
+1. It gets the largest element of array `[3, 1]` which is `3`.
+2. It compares `2` and `3`, `2` is less than `3`.
+
+Therefore, when using descending sort,  `2` is less than `[3, 1]`.
+
+It might seem confusing at first, but note that ascending sort uses the smallest element
+from the array, and descending sort uses the largest element from the array.
+
+For an empty array, it's a special case; an empty array is less than any BSON type.
+In that case, `[]` is less than even `null` for [sorting](#comparison-in-sorting).
+
+## Comparison in Sorting
+
+Sorting used in `.sort()` sorts records by ascending or descending order,
+but array behaves specially.
+
+The following rules are used to compare array and other BSON types.
+
+* Ascending sort uses the smallest element of any BSON type from the array.
+* Descending sort uses the largest element of any BSON type from the array.
+* Empty array is less than null.
+
+### Sorting Array
+
+Let's insert following records to see how arrays are sorted.
+
+```js
+db.mixedsort.insertMany([
+  { _id: 'array-empty', v: [] },
+  { _id: 'null', v: null },
+  { _id: 'array-mixed', v: [ 'a', 'z', 0 ] },
+  { _id: 'array-numbers', v: [ 10, 5 ] },
+  { _id: 'array-strings', v: [ 'a' ] },
+  { _id: 'string', v: 'b' }
+])
+```
+
+Let's use ascending sort on `v`.
+Ascending sort is indicated by `1`.
+
+```js
+db.mixedsort.find().sort({v: 1})
```

Comment:
We should sort by both `{v: 1, _id: 1}` and `{v: 1, _id: -1}` to catch cases where `v` values are considered to equally ordered